std/experimental/logger/asynclogger.d

```diff
@@ -0,0 +1,161 @@
+    {
+        afl.log("test message");
+    }
+
+    auto r = benchmark!(logFileDefault, logFileAsync)(100_000);
+    writeln(map!(a => to!string(to!Duration(a)))(r[]));
+}
+-------------
+*/
+class AsyncLogger : Logger
+{
+    /// Shuntdown message
+    struct ShutdownMsg {}
+
+    import std.datetime : SysTime;
+
+    /** LogEntryMsg is an aggregation copying original $(D LogEntry)
+    but $(D (hasUnsharedAliasing!LogEntryMsg == true)).
+    This aggregation will be send to logging thread.
+    */
+    protected struct LogEntryMsg
+    {
+        /// the filename the log function was called from
+        string file;
+        /// the line number the log function was called from
+        int line;
+        /// the name of the function the log function was called from
+        string funcName;
+        /// the pretty formatted name of the function the log function was
+        /// called from
+        string prettyFuncName;
+        /// the name of the module the log message is coming from
+        string moduleName;
+        /// the $(D LogLevel) associated with the log message
+        LogLevel logLevel;
+        /// the time the message was logged
+        SysTime timestamp;
+        /// the message of the log message
+        string msg;
+    }
+
+    private static void spawnedFunc(shared Logger ptr) @trusted
+    {
+        auto l = cast(Logger)(ptr);
+        bool isRunning = true;
+        while (isRunning)
+        {
+            receive(
+                (LogEntryMsg e)
+                {
+                    LogEntry tmp = LogEntry(e.file, e.line, e.funcName, e.prettyFuncName, e.moduleName, e.logLevel, ownerTid, e.timestamp, e.msg);
+                    l.forwardMsg(tmp);
+                },
+                (ShutdownMsg msg) { isRunning = false; },
+                (Exception ex) { isRunning = false; },
+                (Variant v) {}
+            );
+        }
+    }
+
+    /** This constructor sets a LogLevel and starts a new logging thread
+    that logs every entry asynchronously by using Logger. Because logger writes down
+    all entries into special logging thread, logger should not be used further
+    (without confidence that logger is not being used at the moment).
+    */
+    this(Logger logger, LogLevel lv = LogLevel.all) @trusted
+    {
+        super(lv);
+        _logger = cast(shared Logger)(logger);
+        _tid = spawn(&(spawnedFunc), _logger);
+    }
+
+    /** Returns the Tid of logging thread.
+    It can be useful for setting mailbox size, crowding rules, etc.
+    */
+    Tid tid() @safe
+    {
+        return _tid;
+    }
+
+    override void writeLogMsg(ref LogEntry payload) @trusted
+    {
+        send(_tid, LogEntryMsg(payload.file, payload.line, payload.funcName, payload.prettyFuncName,
+            payload.moduleName, payload.logLevel, payload.timestamp, payload.msg));
+    }
+
+    /// Stops logging thread by sending ShutdownMsg to it.
+    ~this() @trusted
+    {
+        send(_tid, ShutdownMsg());
```

Comment:
Ah right, `OwnerTerminated` should work fine. You didn't explicitly handle it in your receive.